File: internal/config/terms.go
package config

import (
	"fmt"
	"strconv"
	"time"
)

// Term selection should yield smart results based on the current time, as well as the input provided.
// Fall 2024, "spring" => Spring 2025
// Fall 2024, "fall" => Fall 2025
// Summer 2024, "fall" => Fall 2024

const (
	// Fall is the first term of the school year.
	Fall = iota
	// Spring is the second term of the school year.
	Spring
	// Summer is the third term of the school year.
	Summer
)

// Term represents a school term, consisting of a year and a season.
type Term struct {
	Year   uint16
	Season uint8
}

// SeasonRanges represents the start and end day of each term within a year.
type SeasonRanges struct {
	Spring YearDayRange
	Summer YearDayRange
	Fall   YearDayRange
}

// YearDayRange represents the start and end day of a term within a year.
type YearDayRange struct {
	Start uint16
	End   uint16
}

// GetYearDayRange returns the start and end day of each term for the given year.
// The ranges are inclusive of the start day and exclusive of the end day.
func GetYearDayRange(loc *time.Location, year uint16) SeasonRanges {
	springStart := time.Date(int(year), time.January, 14, 0, 0, 0, 0, loc).YearDay()
	springEnd := time.Date(int(year), time.May, 1, 0, 0, 0, 0, loc).YearDay()
	summerStart := time.Date(int(year), time.May, 25, 0, 0, 0, 0, loc).YearDay()
	summerEnd := time.Date(int(year), time.August, 15, 0, 0, 0, 0, loc).YearDay()
	fallStart := time.Date(int(year), time.August, 18, 0, 0, 0, 0, loc).YearDay()
	fallEnd := time.Date(int(year), time.December, 10, 0, 0, 0, 0, loc).YearDay()

	return SeasonRanges{
		Spring: YearDayRange{
			Start: uint16(springStart),
			End:   uint16(springEnd),
		},
		Summer: YearDayRange{
			Start: uint16(summerStart),
			End:   uint16(summerEnd),
		},
		Fall: YearDayRange{
			Start: uint16(fallStart),
			End:   uint16(fallEnd),
		},
	}
}

// GetCurrentTerm returns the current and next terms based on the provided time.
// The current term can be nil if the time falls between terms.
// The 'year' in the term corresponds to the academic year, which may differ from the calendar year.
func GetCurrentTerm(ranges SeasonRanges, now time.Time) (*Term, *Term) {
	literalYear := uint16(now.Year())
	dayOfYear := uint16(now.YearDay())

	// If we're past the end of the summer term, we're 'in' the next school year.
	var termYear uint16
	if dayOfYear > ranges.Summer.End {
		termYear = literalYear + 1
	} else {
		termYear = literalYear
	}

	if (dayOfYear < ranges.Spring.Start) || (dayOfYear >= ranges.Fall.End) {
		// Fall over, Spring not yet begun
		return nil, &Term{Year: termYear, Season: Spring}
	} else if (dayOfYear >= ranges.Spring.Start) && (dayOfYear < ranges.Spring.End) {
		// Spring
		return &Term{Year: termYear, Season: Spring}, &Term{Year: termYear, Season: Summer}
	} else if dayOfYear < ranges.Summer.Start {
		// Spring over, Summer not yet begun
		return nil, &Term{Year: termYear, Season: Summer}
	} else if (dayOfYear >= ranges.Summer.Start) && (dayOfYear < ranges.Summer.End) {
		// Summer
		return &Term{Year: termYear, Season: Summer}, &Term{Year: termYear, Season: Fall}
	} else if dayOfYear < ranges.Fall.Start {
		// Summer over, Fall not yet begun
		return nil, &Term{Year: termYear, Season: Fall}
	} else if (dayOfYear >= ranges.Fall.Start) && (dayOfYear < ranges.Fall.End) {
		// Fall
		return &Term{Year: termYear, Season: Fall}, nil
	}

	panic(fmt.Sprintf("Impossible Code Reached (dayOfYear: %d)", dayOfYear))
}

// ParseTerm converts a Banner term code string to a Term struct.
func ParseTerm(code string) Term {
	year, _ := strconv.ParseUint(code[0:4], 10, 16)

	var season uint8
	termCode := code[4:6]
	switch termCode {
	case "10":
		season = Fall
	case "20":
		season = Spring
	case "30":
		season = Summer
	}

	return Term{
		Year:   uint16(year),
		Season: season,
	}
}

// ToString converts a Term struct to a Banner term code string.
func (term Term) ToString() string {
	var season string
	switch term.Season {
	case Fall:
		season = "10"
	case Spring:
		season = "20"
	case Summer:
		season = "30"
	}

	return fmt.Sprintf("%d%s", term.Year, season)
}

File: src/banner/models/terms.rs
use std::{ops::RangeInclusive, str::FromStr};

use anyhow::Context;
use chrono::{Datelike, Local, NaiveDate};
use serde::{Deserialize, Serialize};

/// The current year at the time of compilation
const CURRENT_YEAR: u32 = compile_time::date!().year() as u32;

/// The valid years for terms
/// We set a semi-static upper limit to avoid having to update this value while also keeping a tight bound
/// TODO: Recheck the lower bound, it's just a guess right now.
const VALID_YEARS: RangeInclusive<u32> = 2007..=(CURRENT_YEAR + 10);

/// Represents a term in the Banner system
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct Term {
    pub year: u32, // 2024, 2025, etc
    pub season: Season,
}

/// Represents the term status at a specific point in time
#[derive(Debug, Clone)]
pub enum TermPoint {
    /// Currently in a term
    InTerm { current: Term },
    /// Between terms, with the next term specified
    BetweenTerms { next: Term },
}

/// Represents a season within a term
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum Season {
    Fall,
    Spring,
    Summer,
}

impl Term {
    /// Returns the current term status - either currently in a term or between terms
    pub fn get_current() -> TermPoint {
        let now = Local::now().naive_local();
        Self::get_status_for_date(now.date())
    }

    /// Returns the current term status for a specific date
    pub fn get_status_for_date(date: NaiveDate) -> TermPoint {
        let literal_year = date.year() as u32;
        let day_of_year = date.ordinal();
        let ranges = Self::get_season_ranges(literal_year);

        // If we're past the end of the summer term, we're 'in' the next school year.
        let term_year = if day_of_year > ranges.summer.end {
            literal_year + 1
        } else {
            literal_year
        };

        if (day_of_year < ranges.spring.start) || (day_of_year >= ranges.fall.end) {
            // Fall over, Spring not yet begun
            TermPoint::BetweenTerms {
                next: Term {
                    year: term_year,
                    season: Season::Spring,
                },
            }
        } else if (day_of_year >= ranges.spring.start) && (day_of_year < ranges.spring.end) {
            // Spring
            TermPoint::InTerm {
                current: Term {
                    year: term_year,
                    season: Season::Spring,
                },
            }
        } else if day_of_year < ranges.summer.start {
            // Spring over, Summer not yet begun
            TermPoint::BetweenTerms {
                next: Term {
                    year: term_year,
                    season: Season::Summer,
                },
            }
        } else if (day_of_year >= ranges.summer.start) && (day_of_year < ranges.summer.end) {
            // Summer
            TermPoint::InTerm {
                current: Term {
                    year: term_year,
                    season: Season::Summer,
                },
            }
        } else if day_of_year < ranges.fall.start {
            // Summer over, Fall not yet begun
            TermPoint::BetweenTerms {
                next: Term {
                    year: term_year,
                    season: Season::Fall,
                },
            }
        } else if (day_of_year >= ranges.fall.start) && (day_of_year < ranges.fall.end) {
            // Fall
            TermPoint::InTerm {
                current: Term {
                    year: term_year,
                    season: Season::Fall,
                },
            }
        } else {
            // This should never happen, but Rust requires exhaustive matching
            panic!("Impossible code reached (dayOfYear: {})", day_of_year);
        }
    }

    /// Returns the start and end day of each term for the given year.
    /// The ranges are inclusive of the start day and exclusive of the end day.
    fn get_season_ranges(year: u32) -> SeasonRanges {
        let spring_start = NaiveDate::from_ymd_opt(year as i32, 1, 14)
            .unwrap()
            .ordinal();
        let spring_end = NaiveDate::from_ymd_opt(year as i32, 5, 1)
            .unwrap()
            .ordinal();
        let summer_start = NaiveDate::from_ymd_opt(year as i32, 5, 25)
            .unwrap()
            .ordinal();
        let summer_end = NaiveDate::from_ymd_opt(year as i32, 8, 15)
            .unwrap()
            .ordinal();
        let fall_start = NaiveDate::from_ymd_opt(year as i32, 8, 18)
            .unwrap()
            .ordinal();
        let fall_end = NaiveDate::from_ymd_opt(year as i32, 12, 10)
            .unwrap()
            .ordinal();

        SeasonRanges {
            spring: YearDayRange {
                start: spring_start,
                end: spring_end,
            },
            summer: YearDayRange {
                start: summer_start,
                end: summer_end,
            },
            fall: YearDayRange {
                start: fall_start,
                end: fall_end,
            },
        }
    }
}

impl TermPoint {
    /// Returns the inner Term regardless of the status
    pub fn inner(&self) -> &Term {
        match self {
            TermPoint::InTerm { current } => current,
            TermPoint::BetweenTerms { next } => next,
        }
    }
}

/// Represents the start and end day of each term within a year
#[derive(Debug, Clone)]
struct SeasonRanges {
    spring: YearDayRange,
    summer: YearDayRange,
    fall: YearDayRange,
}

/// Represents the start and end day of a term within a year
#[derive(Debug, Clone)]
struct YearDayRange {
    start: u32,
    end: u32,
}

impl std::fmt::Display for Term {
    /// Returns the term in the format YYYYXX, where YYYY is the year and XX is the season code
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{year}{season}",
            year = self.year,
            season = self.season.to_str()
        )
    }
}

impl Season {
    /// Returns the season code as a string
    fn to_str(self) -> &'static str {
        match self {
            Season::Fall => "10",
            Season::Spring => "20",
            Season::Summer => "30",
        }
    }
}

impl std::fmt::Display for Season {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Season::Fall => write!(f, "Fall"),
            Season::Spring => write!(f, "Spring"),
            Season::Summer => write!(f, "Summer"),
        }
    }
}

impl FromStr for Season {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let season = match s {
            "10" => Season::Fall,
            "20" => Season::Spring,
            "30" => Season::Summer,
            _ => return Err(anyhow::anyhow!("Invalid season: {s}")),
        };
        Ok(season)
    }
}

impl FromStr for Term {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.len() != 6 {
            return Err(anyhow::anyhow!("Term string must be 6 characters"));
        }

        let year = s[0..4].parse::<u32>().context("Failed to parse year")?;
        if !VALID_YEARS.contains(&year) {
            return Err(anyhow::anyhow!("Year out of range"));
        }

        let season =
            Season::from_str(&s[4..6]).map_err(|e| anyhow::anyhow!("Invalid season: {}", e))?;

        Ok(Term { year, season })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // --- Season::from_str ---

    #[test]
    fn test_season_from_str_fall() {
        assert_eq!(Season::from_str("10").unwrap(), Season::Fall);
    }

    #[test]
    fn test_season_from_str_spring() {
        assert_eq!(Season::from_str("20").unwrap(), Season::Spring);
    }

    #[test]
    fn test_season_from_str_summer() {
        assert_eq!(Season::from_str("30").unwrap(), Season::Summer);
    }

    #[test]
    fn test_season_from_str_invalid() {
        for input in ["00", "40", "1", ""] {
            assert!(
                Season::from_str(input).is_err(),
                "expected Err for {input:?}"
            );
        }
    }

    // --- Season Display ---

    #[test]
    fn test_season_display() {
        assert_eq!(Season::Fall.to_string(), "Fall");
        assert_eq!(Season::Spring.to_string(), "Spring");
        assert_eq!(Season::Summer.to_string(), "Summer");
    }

    #[test]
    fn test_season_to_str_roundtrip() {
        for season in [Season::Fall, Season::Spring, Season::Summer] {
            assert_eq!(Season::from_str(season.to_str()).unwrap(), season);
        }
    }

    // --- Term::from_str ---

    #[test]
    fn test_term_from_str_valid_fall() {
        let term = Term::from_str("202510").unwrap();
        assert_eq!(term.year, 2025);
        assert_eq!(term.season, Season::Fall);
    }

    #[test]
    fn test_term_from_str_valid_spring() {
        let term = Term::from_str("202520").unwrap();
        assert_eq!(term.year, 2025);
        assert_eq!(term.season, Season::Spring);
    }

    #[test]
    fn test_term_from_str_valid_summer() {
        let term = Term::from_str("202530").unwrap();
        assert_eq!(term.year, 2025);
        assert_eq!(term.season, Season::Summer);
    }

    #[test]
    fn test_term_from_str_too_short() {
        assert!(Term::from_str("20251").is_err());
    }

    #[test]
    fn test_term_from_str_too_long() {
        assert!(Term::from_str("2025100").is_err());
    }

    #[test]
    fn test_term_from_str_empty() {
        assert!(Term::from_str("").is_err());
    }

    #[test]
    fn test_term_from_str_invalid_year_chars() {
        assert!(Term::from_str("abcd10").is_err());
    }

    #[test]
    fn test_term_from_str_invalid_season() {
        assert!(Term::from_str("202540").is_err());
    }

    #[test]
    fn test_term_from_str_year_below_range() {
        assert!(Term::from_str("200010").is_err());
    }

    #[test]
    fn test_term_display_roundtrip() {
        for code in ["202510", "202520", "202530"] {
            let term = Term::from_str(code).unwrap();
            assert_eq!(term.to_string(), code);
        }
    }

    // --- Term::get_status_for_date ---

    #[test]
    fn test_status_mid_spring() {
        let date = NaiveDate::from_ymd_opt(2025, 2, 15).unwrap();
        let status = Term::get_status_for_date(date);
        assert!(
            matches!(status, TermPoint::InTerm { current } if current.season == Season::Spring)
        );
    }

    #[test]
    fn test_status_mid_summer() {
        let date = NaiveDate::from_ymd_opt(2025, 7, 1).unwrap();
        let status = Term::get_status_for_date(date);
        assert!(
            matches!(status, TermPoint::InTerm { current } if current.season == Season::Summer)
        );
    }

    #[test]
    fn test_status_mid_fall() {
        let date = NaiveDate::from_ymd_opt(2025, 10, 15).unwrap();
        let status = Term::get_status_for_date(date);
        assert!(matches!(status, TermPoint::InTerm { current } if current.season == Season::Fall));
    }

    #[test]
    fn test_status_between_fall_and_spring() {
        let date = NaiveDate::from_ymd_opt(2025, 1, 1).unwrap();
        let status = Term::get_status_for_date(date);
        assert!(
            matches!(status, TermPoint::BetweenTerms { next } if next.season == Season::Spring)
        );
    }

    #[test]
    fn test_status_between_spring_and_summer() {
        let date = NaiveDate::from_ymd_opt(2025, 5, 15).unwrap();
        let status = Term::get_status_for_date(date);
        assert!(
            matches!(status, TermPoint::BetweenTerms { next } if next.season == Season::Summer)
        );
    }

    #[test]
    fn test_status_between_summer_and_fall() {
        let date = NaiveDate::from_ymd_opt(2025, 8, 16).unwrap();
        let status = Term::get_status_for_date(date);
        assert!(matches!(status, TermPoint::BetweenTerms { next } if next.season == Season::Fall));
    }

    #[test]
    fn test_status_after_fall_end() {
        let date = NaiveDate::from_ymd_opt(2025, 12, 15).unwrap();
        let status = Term::get_status_for_date(date);
        assert!(
            matches!(status, TermPoint::BetweenTerms { next } if next.season == Season::Spring)
        );
        // Year should roll over: fall 2025 ends → next spring is 2026
        let next_term = status.inner();
        assert_eq!(next_term.year, 2026);
    }

    // --- TermPoint::inner ---

    #[test]
    fn test_term_point_inner() {
        let in_term = TermPoint::InTerm {
            current: Term {
                year: 2025,
                season: Season::Fall,
            },
        };
        assert_eq!(
            in_term.inner(),
            &Term {
                year: 2025,
                season: Season::Fall
            }
        );

        let between = TermPoint::BetweenTerms {
            next: Term {
                year: 2026,
                season: Season::Spring,
            },
        };
        assert_eq!(
            between.inner(),
            &Term {
                year: 2026,
                season: Season::Spring
            }
        );
    }
}
